sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java

```diff
@@ -334,94 +373,47 @@ public void setup() {
-      super(connectionConfiguration);
+    @GetInitialRestriction
+    public RedisCursorRange getInitialRestriction() {
+      long dbSize = jedis.dbSize();
+      return RedisCursorRange.of(RedisCursor.of("0", dbSize), RedisCursor.of("0", dbSize));
     }
 
     @ProcessElement
-    public void processElement(ProcessContext c) {
+    public ProcessContinuation processElement(
+        ProcessContext c, RestrictionTracker<RedisCursorRange, RedisCursor> tracker) {
+      RedisCursor cursor = tracker.currentRestriction().getStartPosition();
       ScanParams scanParams = new ScanParams();
       scanParams.match(c.element());
-
-      String cursor = ScanParams.SCAN_POINTER_START;
-      boolean finished = false;
-      while (!finished) {
-        ScanResult<String> scanResult = jedis.scan(cursor, scanParams);
-        List<String> keys = scanResult.getResult();
-        for (String k : keys) {
-          c.output(k);
+      while (tracker.tryClaim(cursor)) {
+        ScanResult<String> scanResult = jedis.scan(cursor.getCursor(), scanParams);
+        List<String> keys = new ArrayList<>();
+        for (String k : scanResult.getResult()) {
+          keys.add(k);
         }
-        cursor = scanResult.getCursor();
-        if (cursor.equals(ScanParams.SCAN_POINTER_START)) {
-          finished = true;
-        }
-      }
-    }
-  }
-
-  /** A {@link DoFn} requesting Redis server to get key/value pairs. */
-  private static class ReadFn extends BaseReadFn<KV<String, String>> {
-    transient @Nullable Multimap<BoundedWindow, String> bundles = null;
-    @Nullable AtomicInteger batchCount = null;
-    private final int batchSize;
-
-    ReadFn(RedisConnectionConfiguration connectionConfiguration, int batchSize) {
-      super(connectionConfiguration);
-      this.batchSize = batchSize;
-    }
-
-    @StartBundle
-    public void startBundle() {
-      bundles = ArrayListMultimap.create();
-      batchCount = new AtomicInteger();
-    }
-
-    @ProcessElement
-    public void processElement(ProcessContext c, BoundedWindow window) {
-      String key = c.element();
-      bundles.put(window, key);
-      if (batchCount.incrementAndGet() > getBatchSize()) {
-        Multimap<BoundedWindow, KV<String, String>> kvs = fetchAndFlush();
-        for (BoundedWindow w : kvs.keySet()) {
-          for (KV<String, String> kv : kvs.get(w)) {
+        if (keys.size() > 0) {
+          for (KV<String, String> kv : fetchAndFlush(keys)) {
             c.output(kv);
           }
         }
-      }
-    }
-
-    @FinishBundle
-    public void finishBundle(FinishBundleContext context) {
-      Multimap<BoundedWindow, KV<String, String>> kvs = fetchAndFlush();
-      for (BoundedWindow w : kvs.keySet()) {
-        for (KV<String, String> kv : kvs.get(w)) {
-          context.output(kv, w.maxTimestamp(), w);
+        if (RedisCursor.ZERO_CURSOR.getCursor().equals(scanResult.getCursor())) {
+          break;
         }
```

Comment:
I would expect the next tryClaim to fail and you shouldn't have to have this exit condition here.

You want the RedisCursor to differentiate `0` as the start or end by embedding a boolean like `isStart`.

Then in the comparison `isStart` compares before everything and `0 && !isStart` compares after everything.